server/routes/experiments/biomed_nl/traversal.py

```diff
@@ -586,21 +593,97 @@ class PathFinder:
         gemini_model_str (str, optional): Name of the Gemini model to use.
     '''
 
+  class QueryTypes(enum.Enum):
+    OVERVIEW = "Overview"
+    TRAVERSAL = "Traversal"
+
+  class DetectedEntities(BaseModel):
+    raw_str: str
+    sanitized_str: str
+    synonyms: list[str]
+
+  class ParseQueryResponse(BaseModel):
+    query_type: "PathFinder.QueryTypes"
+    entities: "list[PathFinder.DetectedEntities]"
+
   def __init__(self,
                query,
-               start_entity_name,
-               start_entity_dcids,
                gemini_client=None,
                gemini_model_str='gemini-2.0-flash-001'):
     # Set params
+    self.raw_query = query
     self.query = query
-    self.start_entity_name = start_entity_name
-    self.start_dcids = start_entity_dcids
+    self.start_entity_name = ''
+    self.start_dcids = []
     self.path_store = PathStore()
     self.input_tokens = 0
     self.output_tokens = 0
     self.gemini = gemini_client
     self.gemini_model_str = gemini_model_str
+    self.query_type = None
+
+  def parse_query(self):
+    '''Parses the raw query using a Gemini model to identify the query type,
+    entities, and their types.
+
+    This method performs the following steps:
+    1. Formats a prompt for the Gemini model using the raw query.
+    2. Calls the Gemini model to generate a structured JSON response containing
+       parsed information about the query.
+    3. Extracts the query type from the Gemini response and sets the
+       `self.query_type` attribute.
+    4. Identifies the primary starting entity from the parsed entities and
+       sets `self.start_entity_name` and creates a list of potential
+       starting strings (including synonyms).
+    5. Creates a mapping between sanitized entity strings and their original
+       raw strings, including synonyms.
+    6. Uses an external function `recognize_entities_from_query` to identify
+       entities and their associated Data Commons IDs (DCIDs) and recognized
+       types within the query.
+    7. Populates the `self.start_dcids` list with the DCIDs of the starting
+       entity.
+    8. Creates a mapping between the raw entity strings and their recognized
+       types.
+    9. Uses an external function `annotate_query_with_types` to add type
+       information to the original raw query, storing the result in
+       `self.query`.
+    '''
+    prompt = utils.PARSE_QUERY_PROMPT.format(QUERY=self.raw_query)
+
+    gemini_response = self.gemini.models.generate_content(
+        model=self.gemini_model_str,
+        contents=prompt,
+        config={
+            'response_mime_type': 'application/json',
+            'response_schema': PathFinder.ParseQueryResponse,
+        })
+    parsed_response = PathFinder.ParseQueryResponse(
```

Comment:
why is parsing the query response going through PathFinder when this response doesn't have to do with path finding. maybe ParseQueryResponse should be a util function

Reply:
discussed offline -- the name ParseQueryResponse was confusing, updated to be more clear